### ISSUE TITLE
All image tags were auto deployed via pipeline when the igor `itemUpperThreshold` changed

### ISSUE THREAD

**Issue by zchee** [spinnaker/spinnaker#2893]
### Issue Summary:
All image tags were deployed(run) via pipeline when the igor `itemUpperThreshold` option changed.

### Cloud Provider(s):
GKE, Kubernetes v1 provider

#### GKE
- Master: `1.8.8-gke.0`
- Nodes: `1.7.5`

#### halyard
`1.2.0-20180509161243`

### Feature Area: 
Pipeline triggers, igor `itemUpperThreshold` config

### Description:
At first, My spinnaker happened following problem. Actually, got `ERROR 1 --- [readScheduler-2] c.n.spinnaker.igor.docker.DockerMonitor  : Number of items (6879) to cache exceeds upper threshold (1000) in ...` error message and _"Not triggering pipeline even if pushed new image to gcr.io"_.

- [Igor threshold blocking GCR triggers for repos w/ > 1000 images #2767](https://github.com/spinnaker/spinnaker/issues/2767)
- [Docker triggers have stopped working #2873](https://github.com/spinnaker/spinnaker/issues/2873)
- [Spinnaker not triggering pipeline when new image is released to Docker registry #2037](https://github.com/spinnaker/spinnaker/issues/2037)

I was digging `clouddriver|igor` pod log, spinnaker issue thread and `clouddriver|igor` sources, and found https://github.com/spinnaker/spinnaker/issues/2873#issuecomment-394966511 comment.
```yaml
spinnaker:
  pollingSafeguard:
    itemUpperThreshold: 50000
```
But I thought it seems to *global* `itemUpperThreshold` when that time.
We don't use `igor`'s jenkins or github features on my spinnaker, and I known `igor` have `dockerRegistry.itemUpperThreshold` and it's valid config from https://github.com/spinnaker/igor/blob/release-1.7.x/igor-web/src/main/groovy/com/netflix/spinnaker/igor/config/DockerRegistryProperties.java#L25

So, I was trying to apply following config to igor in `~/.hal/default/profiles/igor-local.yml` and run `hal deploy apply`:
```yaml
dockerRegistry:
  itemUpperThreshold: 10000
```

---

After that, Spinnaker was auto-deployed via pipeline **all container images(tags) on all applications**.
means triggered `Docker Registry` hook -> `CANCELED` -> triggered-> `SUCCESS` -> triggered -> `CANCELED`. randomly.
In other words, I wanted to running the latest image (let's say `foo:1.0.0` image) on _Foo_ application, run `bar:1.0.0` image on _Bar_ application, But had been deployed and running previous image (e.g., `foo:0.0.1`, `bar:0.1.0`) for a while, caused by above problem.

Is it clouddriver or igor bugs? or, I was some wrong operations?

### Steps to Reproduce:
```console
$ cat ~/.hal/default/profiles/igor-local.yml
```
```yaml
dockerRegistry:
  itemUpperThreshold: 10000
```
```console
$ hal deploy apply
```

### Additional Details:
not yet.

**Comment by lwander** [comment]
Unfortunately, this is what that threshold is for. It's to prevent having large volumes of images trigger pipelines all at once. What you want to do (and sorry this isn't better documented yet) is fastforward those events as described here: https://github.com/spinnaker/spinnaker/issues/2767#issuecomment-388434793. What that does is take all the pending events (that are blocked by that threshold) and store them w/o triggering pipelines.

**Comment by zchee** [comment]
@lwander Thanks for quick reply and advices.

I’ll try to `curl` command solution and check whether the works successfully.

BTW, As I said before, now my spinnaker igor’s have `itemUpperThreshold: 10000`. IIRC it’s default is `1000`.
And, I have huge container image on gcr about 8000~.
In this case, how do handle 7000 images? There is no problem if new(near latest) tags are set to 1000 by spinnaker internal implementation, but if it is reset(refresh), I concerned will happen different problem.

And, How should I order the operations? remove `igor-local.yml` and `hal deploy appy`, than run POST to igor via `curl`. or reverse?

**Comment by tcnksm** [comment]
@lwander 

I also want to make sure decreasing `itemUpperThreshold` number is safe or not ? I worry this operation cause re-indexing and kicked old pipeline again ...

**Comment by lwander** [comment]
Decreasing the threshold shouldn't be a problem -- the reindexing should only happen in three cases:

1. During the upgrade to the release that has this feature, or
2. When a new registry is added, or
3. When your igor cache is deleted

It exists as a safeguard primarily for 2 & 3 to avoid triggering thousands of pipeline executions at once, instead giving you the option to quietly "fast forward" all the tags without triggering any executions.

**Comment by tcnksm** [comment]
Thanks. We're gonna try to decrease the threshold.

**Comment by jaygorrell** [comment]
I just ran into this because of #2 in @lwander list -- isn't that a super-common occurrence? I guess the only time a user wouldn't hit this when setting up Spinnaker is when using a new empty registry? If so, I think the installation documentation is a great place to add coverage.
